core/opendaq/component/src/component_status_container_impl.cpp

```diff
@@ -0,0 +1,97 @@
+#include <opendaq/component_status_container_impl.h>
+#include <coretypes/validation.h>
+#include <opendaq/status_event_args_factory.h>
+
+BEGIN_NAMESPACE_OPENDAQ
+
+ComponentStatusContainerImpl::ComponentStatusContainerImpl()
+    : statuses(Dict<IString, IEnumeration>())
+{
+}
+
+ErrCode ComponentStatusContainerImpl::getStatus(IString* name, IEnumeration** value)
+{
+    OPENDAQ_PARAM_NOT_NULL(name);
+    OPENDAQ_PARAM_NOT_NULL(value);
+
+    std::scoped_lock lock(sync);
+
+    if (!statuses.hasKey(name))
+        return OPENDAQ_ERR_NOTFOUND;
+
+    *value = statuses.get(name).addRefAndReturn();
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode ComponentStatusContainerImpl::getStatuses(IDict** statuses)
+{
+    OPENDAQ_PARAM_NOT_NULL(statuses);
+
+    std::scoped_lock lock(sync);
+
+    *statuses = this->statuses.addRefAndReturn();
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode ComponentStatusContainerImpl::getOnStatusChanged(IEvent** event)
+{
+    OPENDAQ_PARAM_NOT_NULL(event);
+
+    *event = onStatusChangedEvent.addRefAndReturn();
+    return OPENDAQ_SUCCESS;
+}
+
+ErrCode ComponentStatusContainerImpl::addStatus(IString* name, IEnumeration* initialValue)
+{
+    OPENDAQ_PARAM_NOT_NULL(name);
+    OPENDAQ_PARAM_NOT_NULL(initialValue);
+
+    const auto nameObj = StringPtr::Borrow(name);
+    if (nameObj == "")
+        return OPENDAQ_ERR_INVALIDPARAMETER;
+
+    std::scoped_lock lock(sync);
+
+    if (statuses.hasKey(name))
+        return OPENDAQ_ERR_ALREADYEXISTS;
+
+    return statuses->set(name, initialValue);
+}
+
+ErrCode ComponentStatusContainerImpl::setStatus(IString* name, IEnumeration* value)
+{
+    OPENDAQ_PARAM_NOT_NULL(name);
+    OPENDAQ_PARAM_NOT_NULL(value);
+
+    const auto nameObj = StringPtr::Borrow(name);
+    if (nameObj == "")
+        return OPENDAQ_ERR_INVALIDPARAMETER;
+
+    std::scoped_lock lock(sync);
+
+    if (!statuses.hasKey(name))
+        return OPENDAQ_ERR_NOTFOUND;
+
+    const auto valueObj = EnumerationPtr::Borrow(value);
+    const auto oldValue = statuses.get(name);
+    if (valueObj.getEnumerationType() != oldValue.getEnumerationType())
+        return OPENDAQ_ERR_INVALIDTYPE;
+    if (valueObj == oldValue)
+        return OPENDAQ_IGNORED;
+
+    auto errCode = statuses->set(name, value);
+    if (OPENDAQ_FAILED(errCode))
+        return errCode;
+
+    if (onStatusChangedEvent.hasListeners())
```

Comment:
I think we also need to trigger global event on context. Parameters are component and statuseventargs

Reply:
I'm implementing a similar mechanism for `ITags` at the moment. Changes should not be triggered in a local event, but use the global Context one instead (which in turn triggers the `IComponent` changed event).

I opened a PR for the current Tags rework - that can be used for reference. Note that it's still a draft PR.

https://github.com/openDAQ/openDAQ/pull/97